fenicsx: update to v0.4

ma595 opened the pull request:
Renamed Ufcx to FenicsUfcx and added FenicsUfcx dependency to FenicsDolfinx to remove python3 requirement. Updated to latest tagged versions.

chrisrichardson committed:
Fix for xtensor-xsimd

chrisrichardson committed:
Add sha256 for all new releases

ma595 committed:
renamed ufcx package

ma595 committed:
renamed ufcx package

chrisrichardson committed:
Update sha for ffcx

ma595 committed:
fixed hashes and modified fenics-dolfinx to depend on ufcx

ma595 committed:
fixed hashes and modified fenics-dolfinx to depend on ufcx

ma595 committed:
cleaned and fixed dependency types

ma595 committed:
cleaned and fixed dependency types

ma595 committed:
use spec.satisfies in cmake_args

ma595 committed:
use spec.satisfies in cmake_args

chrisrichardson committed:
merge

ma595 committed:
bumped to ufcx@0.4.1

ma595 committed:
bumped to ufcx@0.4.1

ma595 committed:
address PR comments

ma595 committed:
address PR comments

ma595 committed:
fix hashes

ma595 committed:
fix hashes

ma595 committed:
update parmetis in cmake_args to reflect default setting

ma595 committed:
update parmetis in cmake_args to reflect default setting

ma595 committed:
update versions

ma595 committed:
update versions

ma595 committed:
Merge branch 'fenics' into ufcx_0.4.0

chrisrichardson committed:
Merge branch 'develop' into fenics

chrisrichardson committed:
Merge branch 'develop' into ufcx_0.4.0

ma595 committed:
Merge branch 'ufcx_0.4.0' of github.com:FEniCS/spack into ufcx_0.4.0

chrisrichardson committed:
Add dependency fix

garth-wells commented on var/spack/repos/builtin/packages/fenics-basix/package.py:
```diff
@@ -27,6 +29,6 @@ class FenicsBasix(CMakePackage):
     @property
     def root_cmakelists_dir(self):
-        if self.spec.satisfies("@main"):
+        if self.spec.satisfies("@main") or self.spec.satisfies("@0.4.0:"):
```
Is `self.spec.satisfies("@main")` still needed here?

garth-wells commented on var/spack/repos/builtin/packages/fenics-dolfinx/package.py:
```diff
@@ -15,42 +15,45 @@ class FenicsDolfinx(CMakePackage):
     depends_on("mpi")
     depends_on("hdf5+mpi")
     depends_on("boost@1.7.0:+filesystem+program_options+timer")
+    depends_on("parmetis")
```
ParMETIS could be made a variant, defaulting to `True`. What we need is that at least one of (1) ParMETIS, (2) SCOTCH or (3) KaHIP are available. Can this be easily done?

ma595 replied:
I'm not sure how to implement this just yet. Maybe the Spack team can advise?

tldahlgren replied:
One option is through a multi-valued variant (see  https://spack.readthedocs.io/en/latest/packaging_guide.html#multi-valued-variants.)

Or, haven't tried, but you might be able to add a conflict such as:

```
    conflicts("~parmetis~scotch~kahip", msg="At least one of parmetis, scotch, and or kahip must be enabled")
````

garth-wells commented on var/spack/repos/builtin/packages/fenics-dolfinx/package.py:
```diff
@@ -15,42 +15,45 @@ class FenicsDolfinx(CMakePackage):
-    variant("parmetis", default=False, description="parmetis support")
     variant("slepc", default=False, description="slepc support")
     variant("adios2", default=False, description="adios2 support")
+    variant("scotch", default=False, description="scotch support")
```
I think when `scotch` is activated, we need `scotch_mpi`.

ma595 replied:
I thought scotch mpi support would be obtained with the `depends_on("scotch+mpi", when"+scotch")` line 42.

garth-wells commented on var/spack/repos/builtin/packages/fenics-dolfinx/package.py:
```diff
@@ -59,15 +62,15 @@ class FenicsDolfinx(CMakePackage):
     def cmake_args(self):
         args = [
             "-DDOLFINX_SKIP_BUILD_TESTS=True",
+            "-DDOLFINX_ENABLE_PARMETIS=ON",
```
See comments above on ParMETIS.

garth-wells reviewed (changes requested):
I think we can relax the `xtensor` version dependency since we can use xtensor 0.24.

ma595 commented:
> I think we can relax the `xtensor` version dependency since we can use xtensor 0.24.

It looks like xtensor is set to 0.23.10: everywhere. Are you suggesting we bump to 0.24:?

ma595 committed:
bump basix to 0.4.2 and address PR comments

tldahlgren commented on var/spack/repos/builtin/packages/fenics-dolfinx/package.py:
```diff
@@ -59,15 +63,16 @@ class FenicsDolfinx(CMakePackage):
-                'ON' if "+kahip" in self.spec else 'OFF'),
             "-DDOLFINX_ENABLE_PARMETIS=%s" % (
                 'ON' if "+parmetis" in self.spec else 'OFF'),
+            "-DDOLFINX_ENABLE_KAHIP=%s" % (
+                'ON' if "+kahip" in self.spec else 'OFF'),
```
Have you considered using any of the `CMakePackage` helper methods like `define_from_variant()`?  

For more, see https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake.

tldahlgren commented on var/spack/repos/builtin/packages/fenics-dolfinx/package.py:
```diff
@@ -15,12 +15,14 @@ class FenicsDolfinx(CMakePackage):
     maintainers = ["chrisrichardson", "garth-wells", "nate-sime"]
 
     version("main", branch="main")
+    version("0.4.1", sha256="68dcf29a26c750fcea5e02d8d58411e3b054313c3bf6fcbc1d0f08dd2851117f")
```
Is this deliberately `0.4.1` versus `0.4.2` with the other packages?

tldahlgren replied:
Nvm.  Based on changes to other packages, it appears the choice is deliberate.

tldahlgren reviewed (commented):
Confirmed the sha256.

garth-wells committed:
Versioning fixes

garth-wells committed:
Use xtensor-0.24: and loosen pybind11

garth-wells committed:
Add conflicts for partitioners

garth-wells committed:
Updates on partitioners

ma595 committed:
use define_from_variant

tldahlgren commented on var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py:
```diff
@@ -50,8 +50,8 @@ class PyFenicsDolfinx(PythonPackage):
     depends_on("py-petsc4py", type=("build", "run"))
-    depends_on("py-pybind11@2.6.2:2.7", type=("build", "run"))
-    depends_on("xtensor@0.23:", type="build")
+    depends_on("py-pybind11", type=("build", "run"))
```
Are you sure you want to allow pre-`2.6.2` versions?

garth-wells committed:
Tidy up some dependencies

garth-wells committed:
Work on multi-variants for graph partitioners

garth-wells committed:
Fix KaHIP issue.

KaHIP changed the name of its library from 'interface' to 'kahip'. Pin earlier versions of DOLFINx to earlier verisons of KaHIP for proper detection.

garth-wells committed:
Merge branch 'develop' into ufcx_0.4.0

tldahlgren commented:
Was there any intention to add the conflict or is this PR ready to go?

garth-wells commented:
> Was there any intention to add the conflict or is this PR ready to go?

It's ready to go, and has been tested at https://github.com/FEniCS/dolfinx/actions/runs/2346973191.

We did some checking, and the multi-valued variant takes care of the need for at least one partitioner. Spack raises an error for an empty `partitioners=` or `partitioners=abc` if `abc` isn't in the list of  `values`, so the build isalways configured with at least one graph partitioning library.

garth-wells commented:
@tldahlgren many thanks for the very helpful suggestions.

garth-wells committed:
Merge branch 'develop' into ufcx_0.4.0
